### DISCUSSION TITLE
comparison to ajv v8 validator

### DISCUSSION THREAD

**Discussion by lgersman** [colinhacks/zod#3304] [Q&A]
Hi,

I did not find any information about the comparison of the validation features of ajv v8 to zod.

Are there any information about this ? 

Regards,

Lars

**Comment by crrobinson14** [comment]
We've read a lot about `ajv`'s performance so we tried it recently in one of our API stacks. I think the speed is definitely there, but the developer UX is pretty poor. First, compare two schema definitions, the first in AJV:

```
export const announcement = ajv.compile({
  type: 'object',
  properties: {
    client_id: {type: 'string'},
    event_name: {type: 'string'},
    data: {type: 'object'},
    data: {
      type: 'object',
      nullable: true,
      properties: {
        data: {type: 'object'},
        subject: {type: 'string'},
        from: {type: 'string'},
      },
    },
  },
  required: ['client_id', 'event_name', 'data'],
  additionalProperties: false,
} as JSONSchemaType<ICreateAnnouncementRequest>);
```

in Zod:

```
export const CreateAnnouncementSchema = zod.object({
  client_id: zod.string().describe('Client ID'),
  event_name: zod.string().describe('Event name'),
  data: zod
    .object({
      data: zod.any(),
      subject: zod.string(),
      from: zod.string(),
    })
    .optional(),
});

export type TCreateAnnouncementSchema = zod.infer<typeof CreateAnnouncementSchema>;
```

The Zod version is not only more brief, it has some nice-to-haves that AJV can't quite manage, like custom field names and descriptions to drive validation-failure messages (AJV just gives you the "path" to the offending field). For private APIs this may be no big deal but for public API surfaces it's nice to be able to give the developer a bit more of a hint as to what's expected.

The author of ArkType, another validation library like AJV optimized to be as fast as possible, openly states
[in this Reddit post](https://www.reddit.com/r/typescript/comments/17cmt0q/is_zod_actually_that_slow) `I'll tell you a secret- validator performance is mostly negligible in ~95% of situations and Zod's is totally reasonable.` Some apps do need that performance, but if you aren't writing a rela-time equities app processing WebSocket messages at wire-speed, that might not apply to you...
